### DISCUSSION TITLE
FBX skeletal animation/bones and expectations

### DISCUSSION THREAD

**Discussion by ieremius** [assimp/assimp#4781] [Q&A]
I've been attempting to upgrade from AssImp 5.0.1 to 5.2.5 (and a few months ago, I tried 5.2.3, with similar results), and I'm having some problems with FBX skeletal animations in 5.2.5 (and 5.2.3) that I didn't have in 5.0.1.

I can think of three reasons for this...
### Possible Reason 1
Bugs in AssImp. There are some choice Issues entered even recently, for example 
[Bug: Animation data for child bones are completely incorrect.](https://github.com/assimp/assimp/issues/4708)
[Bug: FBX skeleton animation error](https://github.com/assimp/assimp/issues/4620)
In the latter one, Kim K even says "We need to improve our unittests. All tests are green and the bones a broken, really goddamned broken ..."

**That sounds rather conclusive, so we might end the discussion here**!

However, I'm trying to leave no stone unturned before I give up. Possibly my animated FBX models aren't even being affected by these bugs.

### Possible Reason 2
Perhaps there are bugs in our implementation. However, to avoid details of my implementation, we can refer to https://learnopengl.com/Guest-Articles/2020/Skeletal-Animation, which some people will be familiar with, and which our implementation is based on. I see the same results with that demo when I upgrade AssImp. In the comments for that article, there is indication that AssImp 5.0.1 works  with the sample, and some people had to downgrade to that AssImp version to get it to work.

But maybe there is some data from AssImp not taken into account by the LearnOpenGL guest article, which needs to be taken into account post-AssImp 5.0.1. One thing I know of is reading metadata for UpAxis/FrontAxis/CoordAxis, discussed here: [Imported Model is rotated by 90 degrees #849](https://github.com/assimp/assimp/issues/849). However, we've added that. It's not relevant to the issues we're seeing.

Other than that, I'm not aware of any specific bug or limitation in the LearnOpenGL Skeletal Animation guest article, and it's had a lot of review. But I have seen [posts that questioned its veracity](https://gamedev.net/forums/topic/704491-importing-skeleton-from-assimp/), and some [instructions from AssImp committers](https://sourceforge.net/p/assimp/mailman/message/26657996/) which I haven't completely reconciled with the LearnOpenGL guest article. There are also [instructions from previous versions of AssImp](http://sir-kimmi.de/assimp/lib_html/data.html) (under the section "Bones"), which I similarily find hard to reconcile with the guess article, but since they are for previous versions of AssImp I'm not sure I should heed them.

I'm not aware of AssImp official documentation or sample code for handling bones that is specific to AssImp 5/5.2.5.

### Possible Reason 3
Perhaps I have unrealistic expectations. Maybe there is too much variability in the details of FBX exports to expect that say a version of AssImp released in 2022 will continue to work with all kinds of models produced in say 2019?

### Examples
With regards to the specific problems I'm seeing when upgrading from AssImp 5.0.1 to 5.2.5 (or 5.2.3):

IMPORTANTLY, the majority of our FBX models animate fine. So there doesn't seem to be some widespread issue.

Issue 1, surmountable, applies to several of our models: basically like [Imported Model is rotated by 90 degrees #849](https://github.com/assimp/assimp/issues/849), but the fix listed there doesn't address the problem.

Example
5.0.1, correct
![BigDino](https://user-images.githubusercontent.com/77130389/199363900-1d3a7262-78e4-49f8-a5bd-195563d5c0d7.jpg)

5.2.5, incorrect
![BigDino](https://user-images.githubusercontent.com/77130389/199363978-d58bcfc1-0f5f-4f7a-8c01-72b3fcf831cf.jpg)

Issue 2, less surmountable, applies to several of our models: animations are distorted

Examples
5.0.1, correct
![bela](https://user-images.githubusercontent.com/77130389/199364125-7c0de161-1d08-4cec-ae3e-d4fadaad0648.jpg)
![hand](https://user-images.githubusercontent.com/77130389/199364134-bed4326b-80da-4f13-9e27-68e2b050bda9.jpg)

5.2.5, incorrect
![bela](https://user-images.githubusercontent.com/77130389/199364157-64d0ca5a-70d2-47f4-97e4-abe4da656ea1.jpg)
![Hand](https://user-images.githubusercontent.com/77130389/199364169-900b76a1-6ad3-43a4-94ce-d5c19848fdfa.jpg)

(sorry to not show a video, but you get the idea)

### In Conclusion
Sorry for the long post! I'm just trying to decide if
1. It's a simple fact that we're dependent on future AssImp bug fixes
2. There's a bug in our code or equivalently this code https://learnopengl.com/Guest-Articles/2020/Skeletal-Animation
3. I have unrealistic expectations
